### DISCUSSION TITLE
Why does the "is" keyword not narrow the type

### DISCUSSION THREAD

**Discussion by lovetox** [microsoft/pyright#8264] [Q&A] [answered]
```
class A:
    def __init__(self, test: int) -> None:
        pass

class B:
    def __init__(self, test: str) -> None:
        pass

cls_list = {'a': A, 'b': B}

cls_ = cls_list['a']

if cls_ is A:
    cls_(1)
```

This does not narrow the type to A, pyright still thinks `cls_` can be A | B

**Comment by erictraut** [comment]
The type of `cls_` prior to the `if` statement is `A | B`. That means `cls_` could be `A`, `B`, or any subclass of either type.

It's not safe to narrow in this case because `B` is not `final`. There could be another type `C` that is a subclass of both `A` and `B` (and therefore a subclass of `B`). If you mark `B` as `@final`, the narrowing will work as you expect.

**Reply by lovetox** [reply]
I cannot follow your argument.
The `is` keyword tests for identity.
How does it matter that other objects (Subclass of A and B) could exist, when they would not have the same identity as A.
The condition tells me that it must be A, not a subclass. It would evaluate to `False` if i compare `cls_` with any subclass of A or B or both.

**Reply by erictraut** [reply, marked as answer]
Ah, you're correct. Since the RHS is a specific class `A`, as opposed to a variable of type `type[A]`, we can safely narrow the type here. I'll include a fix in the next release.

**Reply by winstxnhdw** [reply]
Has this not yet been fixed? It's been a few releases since.
